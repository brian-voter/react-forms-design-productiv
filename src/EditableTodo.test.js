import React from "react";
import { render, fireEvent } from "@testing-library/react";
import EditableTodo from "./EditableTodo.js";
import { initialTodos } from "./_testCommon.js";
const update = jest.fn();
const remove = jest.fn();

describe("EditableTodo Component smoke and snapshot", function () {
  test("smoke test - renders without crashing", function () {
    render(<EditableTodo todo={initialTodos[0]} update={update} remove={remove} />);
  });

  test("snapshot test", function () {
    const { container } = render(<EditableTodo todo={initialTodos[0]} update={update} remove={remove} />);
    expect(container).toMatchSnapshot();
  });
});

describe("EditableTodo Component functions", function () {
  let editTodo;
  beforeEach(function () {
    editTodo = {
      title: "new title",
      description: "new description",
      priority: 1
    };
  });

  test("Check that update arguments are the form values", function () {
    const update = (todo) => {
      expect(todo.title).toEqual(editTodo.title);
    };

    const { container, queryByText } = render(
      <EditableTodo todo={initialTodos[0]} update={update} remove={remove} />
    );

    expect(queryByText("new title")).not.toBeInTheDocument();
    expect(queryByText(initialTodos[0].title)).toBeInTheDocument();

    fireEvent.click(container.querySelector("#edit-btn"));

    const titleInput = container.querySelector("#newTodo-title");
    const descInput = container.querySelector("#newTodo-description");
    const priorityInput = container.querySelector("#newTodo-priority");

    //Set values for form
    fireEvent.change(titleInput, { target: { value: editTodo.title } });
    fireEvent.change(descInput, { target: { value: editTodo.description } });
    fireEvent.change(priorityInput, { target: { value: editTodo.priority } });

    fireEvent.click(container.querySelector("#submit-btn"));

  });

  test("Check that initial form fields contain existing todo values", function () {
    const { container, queryByText } = render(
      <EditableTodo todo={initialTodos[0]} update={update} remove={remove} />
    );

    expect(queryByText("new title")).not.toBeInTheDocument();
    expect(queryByText(initialTodos[0].title)).toBeInTheDocument();

    fireEvent.click(container.querySelector("#edit-btn"));

    const titleInput = container.querySelector("#newTodo-title");
    const descInput = container.querySelector("#newTodo-description");
    const priorityInput = container.querySelector("#newTodo-priority");

    expect(titleInput.value).toEqual(initialTodos[0].title);
    expect(descInput.value).toEqual(initialTodos[0].description);
    expect(priorityInput.value).toEqual(String(initialTodos[0].priority));

  });

  test("Delete todo", function () {
    const remove = (todoId) => {
      expect(todoId).toEqual(initialTodos[0].id);
    };

    const { container, queryByText } = render(
      <EditableTodo todo={initialTodos[0]} update={update} remove={remove} />
    );

    expect(queryByText(initialTodos[0].title)).toBeInTheDocument();

    fireEvent.click(container.querySelector("#delete-btn"));
  });

});