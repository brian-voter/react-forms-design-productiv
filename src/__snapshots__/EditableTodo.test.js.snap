// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditableTodo Component smoke and snapshot snapshot test 1`] = `
<div>
  <div
    class="EditableTodo"
  >
    <div
      class="mb-3"
    >
      <div
        class="float-end text-sm-end"
      >
        <button
          class="EditableTodo-toggle btn-link btn btn-sm"
          id="edit-btn"
        >
          Edit
        </button>
        <button
          class="EditableTodo-delBtn btn-link btn btn-sm text-danger"
          id="delete-btn"
        >
          Del
        </button>
      </div>
      <div
        class="Todo"
      >
        <div>
          <b>
            Code!
          </b>
           
          <small>
            (priority: 
            2
            )
          </small>
        </div>
        <div>
          <small>
            Write some code
          </small>
        </div>
      </div>
    </div>
  </div>
</div>
`;
